lib/tapp.hoon

```diff
@@ -0,0 +1,337 @@
+    ==
+|%
+++  tapp-sur  (^tapp-sur out-poke-data out-peer-data)
+++  card  card:tapp-sur
+++  sign  sign:tapp-sur
+++  contract  contract:tapp-sur
+++  command
+  $%  [%poke =in-poke-data]
+      [%peer =path]
+      [%diff =dock =path =in-peer-data]
+      [%take =sign]
+  ==
+::
+++  trad-lib  (^trad sign card contract)
+++  trad  trad:trad-lib
+::
++$  move  (pair bone card)
+++  tapp-trad  (trad state-type)
++$  tapp-state
+  $:  waiting=(qeu command)
+      active=(unit eval-form:eval:tapp-trad)
+      app-state=state-type
+  ==
+::
+::  The form of a tapp that only handles pokes
+::
+++  tapp-core-poke
+  $_  ^|
+  |_  [bowl:gall state-type]
+  ++  handle-poke
+    |~  in-poke-data
+    *form:tapp-trad
+  --
+::
+++  create-tapp-poke
+  |=  handler=tapp-core-poke
+  %-  create-tapp-poke-peer
+  |_  [=bowl:gall state=state-type]
+  ++  handle-poke  ~(handle-poke handler bowl state)
+  ++  handle-peer  |=(* (trad-fail:trad-lib %no-peer-handler >path< ~))
+  --
+::
+::  The form of a tapp that only handles pokes and peers
+::
+++  tapp-core-poke-peer
+  $_  ^|
+  |_  [bowl:gall state-type]
+  ++  handle-poke
+    |~  in-poke-data
+    *form:tapp-trad
+  ::
+  ++  handle-peer
+    |~  path
+    *form:tapp-trad
+  --
+::
+++  create-tapp-poke-peer
+  |=  handler=tapp-core-poke-peer
+  %-  create-tapp-all
+  |_  [=bowl:gall state=state-type]
+  ++  handle-poke  ~(handle-poke handler bowl state)
+  ++  handle-peer  ~(handle-peer handler bowl state)
+  ++  handle-diff  |=(* (trad-fail:trad-lib %no-diff-handler >path< ~))
+  ++  handle-take  |=(* (trad-fail:trad-lib %no-take-handler >path< ~))
+  --
+::
+::  The form of a tapp that only handles pokes and diffs
+::
+++  tapp-core-poke-diff
+  $_  ^|
+  |_  [bowl:gall state-type]
+  ++  handle-poke
+    |~  in-poke-data
+    *form:tapp-trad
+  ::
+  ++  handle-diff
+    |~  [dock path in-peer-data]
+    *form:tapp-trad
+  --
+::
+++  create-tapp-poke-diff
+  |=  handler=tapp-core-poke-diff
+  %-  create-tapp-all
+  |_  [=bowl:gall state=state-type]
+  ++  handle-poke  ~(handle-poke handler bowl state)
+  ++  handle-peer  |=(* (trad-fail:trad-lib %no-peer-handler >path< ~))
+  ++  handle-diff  ~(handle-diff handler bowl state)
+  ++  handle-take  |=(* (trad-fail:trad-lib %no-take-handler >path< ~))
+  --
+::
+::  The form of a tapp that only handles pokes, peers, and takes
+::
+++  tapp-core-poke-peer-take
+  $_  ^|
+  |_  [bowl:gall state-type]
+  ++  handle-poke
+    |~  in-poke-data
+    *form:tapp-trad
+  ::
+  ++  handle-peer
+    |~  path
+    *form:tapp-trad
+  ::
+  ++  handle-take
+    |~  sign
+    *form:tapp-trad
+  --
+::
+++  create-tapp-poke-peer-take
+  |=  handler=tapp-core-poke-peer-take
+  %-  create-tapp-all
+  |_  [=bowl:gall state=state-type]
+  ++  handle-poke  ~(handle-poke handler bowl state)
+  ++  handle-peer  ~(handle-peer handler bowl state)
+  ++  handle-diff  |=(* (trad-fail:trad-lib %no-diff-handler >path< ~))
+  ++  handle-take  ~(handle-take handler bowl state)
+  --
```

Comment:
The repetition of all of the above is a bit meh, but I don't think we have any fancy affordances that could make this better? It's fine as-is, but already shows that it scales pretty poorly. Worth taking note of.

Reply:
We don't have any easy way to say "call this if it exists", and I don't think the `core` type is flexible enough to make that work.  The other option would be to say that everyone has to implement all the arms in `tapp-core-all`.  You could say that each arm is either the gate or null, so not handling `diff`, `take`, and peer would look like this at the bottom of the app:

```
++  handle-diff  ~
++  handle-take  ~
++  handle-peer  ~
```